versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/DatabaseAdapter.java

```diff
@@ -291,6 +293,56 @@ void updateRepositoryDescription(Function<RepoDescription, RepoDescription> upda
    */
   Stream<CommitLogEntry> scanAllCommitLogEntries();
 
+  /**
+   * Retrieve the known attachment keys for a content ID.
+   *
+   * <p>Implementations may return keys that do not or no longer exist. Especially non-transactional
+   * database adapter implementations allow this as a compromise. See {@link
+   * #mapToAttachment(Stream)}, which only returns existing content attachments.
+   */
+  Stream<ContentAttachmentKey> getAttachmentKeys(String contentId);
+
+  /**
+   * Retrieve the content attachments identified via {@code keys}. Attachments that do not exist are
+   * not returned.
+   *
+   * <p>Whether the input stream is "terminated" or "fluently" mapped, can vary between
+   * implementations.
+   */
+  Stream<ContentAttachment> mapToAttachment(Stream<ContentAttachmentKey> keys);
+
+  /**
+   * Consistent put-attachment operation.
+   *
+   * <p>Either a "put-if-absent", if {@code expectedHash} is empty or a compare-and-swap based on
+   * the value of {@code expectedVersion}.
+   *
+   * <p>Note: this method uses conditional put operations, unlike {@link #putAttachments(Stream)}.
+   * Users should not use this method and {@link #putAttachments(Stream)} for the same keys.
+   *
+   * @param attachment the attachment to write
+   * @param expectedVersion indicator for put-if-absent or the expected value on an existing item
+   */
+  boolean consistentPutAttachment(ContentAttachment attachment, Optional<String> expectedVersion);
```

Comment:
Just wondering if we may need change log for attachments? Since the key is probably going to be frozen into a commit, could it be necessary for us to get previous versions of attachments when/after changes are made? Since we track top-level content changes in commits, should we also track attachment changes for the same of having full history?

Reply:
Not sure TBH. We do not need this consisten-put at the moment. It was intended to potentially allow sequence-numbers per content-id - like column-IDs and such, if we want to at some point.